File: tinysat-java/src/tinysat/base/DimacsParser.java
package tinysat.base;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

public class DimacsParser {
    public static Instance parseProgram(String program) {
        List<String> lines = Arrays.asList(program.split("\n"));

        // Find the first line with the format: p <varCount> <clauseCount>
        String current = "";
        Iterator<String> iter = lines.iterator();
        while (!current.startsWith("p") && lines.iterator().hasNext()) {
            current = iter.next();
        }
        // TODO: Throw error if !current.startsWith("p")

        String[] tokens = current.split("\\s+");
        int varCount = Integer.parseInt(tokens[2]);
        int clauseCount = Integer.parseInt(tokens[3]);
        List<List<Integer>> clauses = new ArrayList<>();

        while (iter.hasNext() && clauseCount > 0) {
            --clauseCount;
            current = iter.next();
            // TODO: Throw error if the line doesn't end with '0'

            tokens = current.split("\\s+");
            List<Integer> clause = new ArrayList<>();
            for (int i = 0; i < tokens.length - 1; ++i) {
                clause.add(encodeLiteral(Integer.parseInt(tokens[i])));
            }
            clauses.add(clause);
        }

        Instance ins = new Instance(varCount, clauses);
        return ins;
    }

    public static int[] decodeAssignments(List<Variable> assignment) {
        int[] result = new int[assignment.size()];
        for (int i = 0; i < assignment.size(); ++i) {
            result[i] = decodeVariable(assignment.get(i));
        }
        return result;
    }

    private static int encodeLiteral(int x) {
        if (x > 0) {
            return (x - 1) * 2;
        }
        return (-x - 1) * 2 + 1;
    }

    private static int decodeVariable(Variable var) {
        int x = var.getId();
        if (!var.getValue())
            x = -x;
        return x;
    }
}

File: tinysat-python/base/dimacs_parser.py
from base.instance import Instance

"""
Parse Input/Output in DIMACS format.
"""


def __encode_literal(x):
    return (x-1) * 2 if x > 0 else (-x - 1) * 2 + 1


def __parse_clause(line):
    """
    Converting a clause to an array of literals.
    """        
    literals = [int(x) for x in line.split()]
    if literals[-1] != 0:
        raise Exception("Parsing error: All clauses must end with 0.")
    
    literals = [__encode_literal(x) for x in literals[:-1]]
    return literals


def parse_program(program):
    """
    Parse a program (of type string) and return an Instance
    """
    # TODO: Check number of vars and clauses to see if they match

    lines = program.split("\n")

    # Ignore comments at the beginning of the file:
    start_i = 0
    while lines[start_i][0] != "p":
        start_i += 1
    lines = lines[start_i:]

    var_count, clause_count = lines[0].split()[2:4]
    var_count, clause_count = int(var_count), int(clause_count)

    variables = list(range(var_count))
    clauses = []
    for i in range(1, clause_count + 1):
        clauses.append(__parse_clause(lines[i]))

    return Instance(variables, clauses)


def decode_assignment(assignment):
    result = [i + 1 if assignment[i] else -(i + 1) for i in range(len(assignment))]
    return " ".join([str(x) for x in result])
